### PR TITLE
Process logs

### PR DESCRIPTION
**Pull request by VaskoBozhurski**
#### Description: 

Instead of updating log entries on each step, the process log service will
now save each step as separate file and merge them only on demand when the
content is being requested.

This commit also changes back the column for file content for the process_log
table back to bytea.

#### Issue: 
LMCROSSITXSADEPLOY-1424

### CODE REVIEW

File: com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/query/providers/SqlFileQueryProvider.java [hunk 1]
```diff
@@ -63,7 +63,11 @@ public SqlFileQueryProvider(String tableName, DataSourceDialect dataSourceDialec
                 statement.setString(7, fileEntry.getDigestAlgorithm());
                 statement.setTimestamp(8, new Timestamp(fileEntry.getModified()
                     .getTime()));
-                getDataSourceDialect().setBlobAsBinaryStream(statement, 9, content);
+                if (isContentStoredAsBlob) {
+                    getDataSourceDialect().setBlobAsBinaryStream(statement, 9, content);
```
Comment (petko-boyadzhiev) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/query/providers/SqlFileQueryProvider.java]: Make the class abstract with the two abstract methods for getting and storing content. Use one of them in place of this if statement.

File: com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/query/providers/SqlFileQueryProvider.java [hunk 2]
```diff
@@ -176,7 +180,7 @@ public SqlFileQueryProvider(String tableName, DataSourceDialect dataSourceDialec
         };
     }
 
-    public SqlQuery<Void> getProcessFileWithContentQuery(final FileDownloadProcessor fileDownloadProcessor) {
+    public SqlQuery<Void> getProcessFileWithContentQuery(final FileDownloadProcessor fileDownloadProcessor, boolean isContentStoredAsBlob) {
```
Comment (petko-boyadzhiev) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/query/providers/SqlFileQueryProvider.java]: Remove boolean. Use the abstract class approach from above.

Comment (petko-boyadzhiev) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/query/providers/SqlFileQueryProvider.java]: Remove boolean. Use the abstract class approach from above.

File: com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/query/providers/SqlFileQueryProvider.java [hunk 3]
```diff
@@ -188,7 +192,7 @@ public SqlFileQueryProvider(String tableName, DataSourceDialect dataSourceDialec
                     .getSpace());
                 resultSet = statement.executeQuery();
                 if (resultSet.next()) {
-                    processFileContent(resultSet, fileDownloadProcessor);
+                    processFileContent(resultSet, fileDownloadProcessor, isContentStoredAsBlob);
```
Comment (petko-boyadzhiev) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/query/providers/SqlFileQueryProvider.java]: Remove boolean. Use the abstract class approach from above.

File: com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/query/providers/SqlFileQueryProvider.java [hunk 4]
```diff
@@ -328,8 +332,14 @@ private DataSourceDialect getDataSourceDialect() {
         return dataSourceDialect;
     }
 
-    private void processFileContent(ResultSet resultSet, final FileDownloadProcessor fileDownloadProcessor) throws SQLException {
-        InputStream fileStream = getDataSourceDialect().getBinaryStreamFromBlob(resultSet, FileServiceColumnNames.CONTENT);
+    private void processFileContent(ResultSet resultSet, final FileDownloadProcessor fileDownloadProcessor, boolean isContentStoredAsBlob)
+        throws SQLException {
+        InputStream fileStream;
+        if (isContentStoredAsBlob) {
+            fileStream = getDataSourceDialect().getBinaryStreamFromBlob(resultSet, FileServiceColumnNames.CONTENT);
```
Comment (petko-boyadzhiev) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/query/providers/SqlFileQueryProvider.java]: Remove if. Use the abstract class approach.

File: com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/services/ProcessLogsPersistenceService.java [hunk 5]
```diff
@@ -50,66 +48,49 @@ public ProcessLogsPersistenceService(DataSourceWithDialect dataSourceWithDialect
...
-    }
-
     private void storeLogFile(final String space, final String namespace, final String remoteLogName, File localLog)
         throws FileStorageException {
         try (InputStream inputStream = new FileInputStream(localLog)) {
             FileEntry localLogFileEntry = createFileEntry(space, namespace, remoteLogName, localLog);
-
-            getSqlQueryExecutor().execute(getSqlFileQueryProvider().getStoreFileQuery(localLogFileEntry, inputStream));
+            getSqlQueryExecutor()
+                .execute(getSqlFileQueryProvider().getStoreFileQuery(localLogFileEntry, inputStream, isContentStoredAsBlob));
```
Comment (petko-boyadzhiev) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/services/ProcessLogsPersistenceService.java]: Remove boolean. Use the abstract class approach from above.

File: com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/services/ProcessLogsPersistenceService.java [hunk 6]
```diff
@@ -50,66 +48,49 @@ public ProcessLogsPersistenceService(DataSourceWithDialect dataSourceWithDialect
 
     public String getLogContent(String space, String namespace, String logName) throws FileStorageException {
         final StringBuilder builder = new StringBuilder();
-        String logId = findFileId(space, namespace, logName);
+        List<String> logIds = findAndSortFileIds(space, namespace, logName);
```
Comment (petko-boyadzhiev) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/services/ProcessLogsPersistenceService.java]: Either replace the method with the lambda or rename it to something like getFileIds

File: com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-alter_process_log_alter_content_column_to_bytea_cf.xml [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog/1.9
+                      http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-1.9.xsd">
+
+    <changeSet author="i328376" id="alter_process_log_alter_content_column_to_bytea_cf">
+        <preConditions onFail="MARK_RAN">
+            <dbms type="postgresql" />
+        </preConditions>
+        <delete tableName="process_log" />
+        <sql>
+            DROP TRIGGER IF EXISTS delete_from_system_table_process_log on process_log;
```
Comment (petko-boyadzhiev) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-alter_process_log_alter_content_column_to_bytea_cf.xml]: What about delete_from_system_table_lm_sl_persistance_file?

Comment (VaskoBozhurski) [reply com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-alter_process_log_alter_content_column_to_bytea_cf.xml]: lm_sl_persistance_file still uses blob and will still need to have the trigger to ensure proper cleanup of blobs on delete.

File: com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/services/ProcessLogsPersistenceService.java [hunk 8]
```diff
@@ -50,65 +53,42 @@ public ProcessLogsPersistenceService(DataSourceWithDialect dataSourceWithDialect
...
         return builder.toString();
     }
 
-    public boolean deleteLog(String space, String namespace, String logName) throws FileStorageException {
-        String fileId = findFileId(space, namespace, logName);
-        return deleteFile(space, fileId);
-    }
-
-    private String findFileId(String space, String namespace, String fileName) throws FileStorageException {
+    private List<String> getSortedFileIds(String space, String namespace, String fileName) throws FileStorageException {
```
Comment (petko-boyadzhiev) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/java/com/sap/cloud/lm/sl/cf/persistence/services/ProcessLogsPersistenceService.java]: Extract the sorting outside

File: com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-remove_process_log_unique_constraint.xml [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog
+    logicalFilePath="com/sap/cloud/lm/sl/persistence/db/changelog/db-changelog-add_process_log_table.xml"
+    xmlns="http://www.liquibase.org/xml/ns/dbchangelog/1.9" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog/1.9
+                      http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-1.9.xsd">
+    <changeSet author="i328376" id="remove_process_log_unique_constrain">
```
Comment (boyan-velinov) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-remove_process_log_unique_constraint.xml]: Remove i-user. Use "sap.com"

File: com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-remove_process_log_unique_constraint.xml [hunk 10]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (boyan-velinov) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-remove_process_log_unique_constraint.xml]: Did you test in on both Postgre and HANA?

File: com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-remove_process_log_unique_constraint.xml [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+<databaseChangeLog
+    logicalFilePath="com/sap/cloud/lm/sl/persistence/db/changelog/db-changelog-add_process_log_table.xml"
+    xmlns="http://www.liquibase.org/xml/ns/dbchangelog/1.9" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog/1.9
+                      http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-1.9.xsd">
+    <changeSet author="i328376" id="remove_process_log_unique_constrain">
+        <dropUniqueConstraint constraintName="unq_process_log" tableName="process_log" />
```
Comment (boyan-velinov) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-remove_process_log_unique_constraint.xml]: Probably some precondition:
`<preConditions onFail="MARK_RAN">
    <tableExists tableName="process_log" />
</preConditions>`

File: com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-alter_process_log_alter_content_column_to_bytea_cf.xml [hunk 12]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog
+    logicalFilePath="com/sap/cloud/lm/sl/persistence/db/changelog/db-changelog-add_trigger_for_lm_sl_persistance_file_process_log_cf.xml"
+    xmlns="http://www.liquibase.org/xml/ns/dbchangelog/1.9" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog/1.9
+                      http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-1.9.xsd">
+
+    <changeSet author="i328376" id="alter_process_log_alter_content_column_to_bytea_cf">
```
Comment (boyan-velinov) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-alter_process_log_alter_content_column_to_bytea_cf.xml]: remove i-user

File: com.sap.cloud.lm.sl.cf.web/src/main/webapp/WEB-INF/services-context.xml [hunk 13]
```diff
@@ -24,6 +24,13 @@
     </bean>
 
     <beans profile="cf">
+        <bean id="processLogsPersistenceService"
+            class="com.sap.cloud.lm.sl.cf.persistence.services.ProcessLogsPersistenceService">
+            <constructor-arg name="dataSourceWithDialect" ref="dataSourceWithDialect" />
+            <constructor-arg name="isContentStoredAsBlob">
```
Comment (boyan-velinov) [inline com.sap.cloud.lm.sl.cf.web/src/main/webapp/WEB-INF/services-context.xml]: can be done in 1 line

File: com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-alter_process_log_alter_content_column_to_bytea_cf.xml [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+        <preConditions onFail="MARK_RAN">
+            <dbms type="postgresql" />
+        </preConditions>
+        <delete tableName="process_log" />
+        <sql>
```
Comment (boyan-velinov) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-alter_process_log_alter_content_column_to_bytea_cf.xml]: Can we add precondition for existence of table "process_log"

File: com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-alter_process_log_alter_content_column_to_bytea_cf.xml [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+    logicalFilePath="com/sap/cloud/lm/sl/persistence/db/changelog/db-changelog-add_trigger_for_lm_sl_persistance_file_process_log_cf.xml"
+    xmlns="http://www.liquibase.org/xml/ns/dbchangelog/1.9" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog/1.9
+                      http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-1.9.xsd">
+
+    <changeSet author="i328376" id="alter_process_log_alter_content_column_to_bytea_cf">
+        <preConditions onFail="MARK_RAN">
+            <dbms type="postgresql" />
+        </preConditions>
+        <delete tableName="process_log" />
```
Comment (boyan-velinov) [inline com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-alter_process_log_alter_content_column_to_bytea_cf.xml]: Why delete whole table process_log and then drop column from it.
I guess this is leftover?

Comment (VaskoBozhurski) [reply com.sap.cloud.lm.sl.cf.persistence/src/main/resources/com/sap/cloud/lm/sl/cf/persistence/db/changelog/db-changelog-alter_process_log_alter_content_column_to_bytea_cf.xml]: delete actually executes DELETE FROM process_log; and not drop table :)

**Pull request merged**